pubsub/google/cloud/pubsub_v1/subscriber/_helper_threads.py

```diff
@@ -42,27 +75,43 @@ class QueueCallbackWorker(object):
             concurrency boundary implemented by ``executor``. Items will
             be popped off (with a blocking ``get()``) until :attr:`STOP`
             is encountered.
-        callback (Callable[[str, Dict], Any]): A callback that can process
-            items pulled off of the queue. Items are assumed to be a pair
-            of a method name to be invoked and a dictionary of keyword
-            arguments for that method.
+        callback (Callable[Sequence[Any], Any]): A callback that can process
+            items pulled off of the queue. Multiple items will be passed to
+            the callback in batches.
+        max_items (int): The maximum amount of items that will be passed to the
+            callback at a time.
+        max_latency (float): The maximum amount of time in seconds to wait for
+            additional items before executing the callback.
     """
 
-    def __init__(self, queue, callback):
+    def __init__(self, queue, callback, max_items=100, max_latency=0):
         self.queue = queue
         self._callback = callback
+        self.max_items = max_items
+        self.max_latency = max_latency
 
     def __call__(self):
-        while True:
-            item = self.queue.get()
-            if item == STOP:
-                _LOGGER.debug('Exiting the QueueCallbackWorker.')
-                return
+        should_stop = False
+        while not should_stop:
+            items = _get_many(
+                self.queue,
+                max_items=self.max_items,
+                max_latency=self.max_latency)
+
+            # If stop is in the items, process all items up to STOP and then
+            # exit.
+            should_stop = False
+            try:
+                items = items[:items.index(STOP)]
```

Comment:
if/else better in this case?

Reply:
Eh, it wouldn't be quite as tidy. You could do:

```
if STOP in items:
    cutoff = items.index(STOP)
    items = items[:items.index(STOP)
```

But that incurs two operations on the list (once to find if stop is there, another to find where stop is).

Reply:
I was thinking more of likely cases and branch prediction issues, but probably microoptimizing (whether STOP is more likely than not to happen)